### PR TITLE
Add ESM distributions for browser

### LINKED ISSUE #8772: Native ESM support
We have documents for ESM import on browser but it needs using some module bundler (e.g. webpack, rollup). (See https://prettier.io/docs/en/browser.html#es-modules)
I thinks we should have native ESM standalone bundled. I hope we can use Prettier like below:
```js
import prettier from "https://unpkg.com/prettier@2.0.5/standalone.m.js";
import parserGraphql from "https://unpkg.com/prettier@2.0.5/parser-graphql.m.js";

prettier.format("query { }", {
  parser: "graphql",
  plugins: [parserGraphql],
});
```
WDYT?

### PR DESCRIPTION
**Pull request by Monchi**
This PR adds configurations to build ESM version of universal bundles. A part of #8772.

Since webpack 4 cannot output ESM, the package built with webpack (`parser-postcss`) is not supported in this PR.

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Ready?

**Comment by Monchi** [comment]
@fisker Yes

File: package.json
Comment (fisker) [inline package.json:23]: I don't think we want this `exports`

File: scripts/build/config.js
Comment (fisker) [inline scripts/build/config.js:161]: Should be `.mjs`.

FYI: I saw you post the reason in `babel` project, but do we have the same problem?

File: changelog_unreleased/api/pr-8983.md
Comment (fisker) [inline changelog_unreleased/api/pr-8983.md:13]: This can't work in browser, should be something like 


```js
import prettier from "https://unpkg.com/prettier@2.0.5/standalone.mjs";
import parserGraphql from "https://unpkg.com/prettier@2.0.5//parser-graphql.mjs";

prettier.format("query {     }", {
  parser: "graphql",
  plugins: [parserGraphql],
});
```

**Comment by fisker** [comment]
> Since webpack 4 cannot output ESM, the package built with webpack (parser-postcss) is not supported in this PR.

Can we bundle `parser-postcss.js` with webpack first, and build `parser-postcss.mjs` from bundled `parser-postcss.js`?

We definitely don't want ship the esm support without `parser-postcss.js`

File: changelog_unreleased/api/pr-8983.md
Comment (Monchi) [reply changelog_unreleased/api/pr-8983.md:13]: Ahh yes, I didn't mean that this code works on browser, but it's misleading. Fixed.

File: package.json
Comment (Monchi) [reply package.json:23]: @fisker I think [current documented example](https://prettier.io/docs/en/browser.html#es-modules) will be broken without `exports`.

I'd remove `exports` if you want even though it's breaking change.

File: scripts/build/config.js
Comment (Monchi) [reply scripts/build/config.js:161]: @fisker 

> FYI: I saw you post the reason in babel project, but do we have the same problem?

Yes. It seems that webpack 4 resolve `*.js` as CommonJS when there is any `*.mjs`. This will break `*.js` written as ESM. So we need to use `.module.js` instead of `.mjs` for now.

**Comment by Monchi** [comment]
> Can we bundle parser-postcss.js with webpack first, and build parser-postcss.mjs from bundled parser-postcss.js?

Good idea, thanks! I'll try it.

File: scripts/build/config.js
Comment (fisker) [reply scripts/build/config.js:161]: Sorry, I don't understand this, why we should care about webpack?

Comment (Monchi) [reply scripts/build/config.js:161]: I'm sorry about not enough explanation. This is not a problem about Prettier's bundler.

A package that bundles Prettier as ESM with webpack would face this problem. Bundling the code below may cause an error because webpack wrongly resolves ESM with `.js` extension which prettier depends on as CJS.

```js
import prettier from "prettier/standalone.mjs";
import parserGraphql from "prettier/parser-graphql.mjs";

prettier.format("query {     }", {
  parser: "graphql",
  plugins: [parserGraphql],
});
```

**Comment by Monchi** [comment]
Re-running CI because it failed due to registry's error.

```
error An unexpected error occurred: "https://registry.yarnpkg.com/@rollup/plugin-json/-/plugin-json-4.1.0.tgz: Request failed \"503 Service Unavailable\"".
info If you think this is a bug, please open a bug report with the information provided in "/home/runner/work/prettier/prettier/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

File: scripts/build/config.js
Comment (evilebottnawi) [reply scripts/build/config.js:170]: > A package that bundles Prettier as ESM with webpack would face this problem. Bundling the code below may cause an error because webpack wrongly resolves ESM with .js extension which prettier depends on as CJS.

No, using `import` and `export` automatically enable ESM

Comment (evilebottnawi) [reply scripts/build/config.js:170]: I don't understand problem with webpack here, despite the fact that I am his contributor

File: package.json
Comment (evilebottnawi) [inline package.json:71]: It is breaking change, and can't be here

File: scripts/build/config.js
Comment (Monchi) [reply scripts/build/config.js:170]: @evilebottnawi Thank you for reviewing! I'm not an expert of webpack/ESM and maybe misunderstanding.

So [this comment](https://github.com/reactjs/react-transition-group/pull/77#issuecomment-453940955) is not true?
I referred these workarounds: https://github.com/preactjs/preact/pull/1425, https://github.com/formatjs/formatjs/issues/1395

Comment (Monchi) [reply scripts/build/config.js]: @evilebottnawi @fisker Sorry, I noticed two things.

1. Prettier's standalone bundle doesn't resolve any dependencies. So we don't need to care about the webpack's behavior.

2. `exports` field is breaking change. Should revert it.

Applied these things in 7b02e1d and 2762bbf

File: package.json
Comment (fisker) [inline package.json:13]: Having both `standalone.js` and `standalone.mjs` at the same location may cause problem when import without extensions, maybe we can put it in `esm/standalone.mjs`?

File: changelog_unreleased/api/pr-8983.md
Comment (fisker) [inline changelog_unreleased/api/pr-8983.md:7]: I don't think we can omit the extensions in browser, should use `https://unpkg.com/prettier/parser-graphql.mjs`.

**Comment by fisker** [comment]
`parser-flow.js` not working, it's not in strict, maybe we should merge #8311

**Comment by Monchi** [comment]
@fisker Exactly! `parser-flow` works after merging #8311. Let's merge this PR after #8311.

**Comment by fisker** [comment]
#8311 merged, we should add some simple tests for this.

File: scripts/build/bundler.js
Comment (ExE-Boss) [inline scripts/build/bundler.js:226]: I’d prefer if this was done using `dist/esm/package.json` with content <code>{ "[type][package-json-type-field]": "module" }</code> and:

```suggestion
          file: `dist/esm/${bundle.output}`,
```

[package-json-type-field]: https://nodejs.org/api/esm.html#esm_package_json_type_field

Comment (fisker) [reply scripts/build/bundler.js:226]: We are making if work in browser, not node.js.

Comment (ExE-Boss) [reply scripts/build/bundler.js:226]: And browsers prefer the `.js` file extension.

Comment (fisker) [reply scripts/build/bundler.js:226]: Even for npm package, I prefer use `.mjs` extension.

Comment (fisker) [reply scripts/build/bundler.js:226]: > And browsers prefer the `.js` file extension.

Who says that?

Comment (fisker) [reply scripts/build/bundler.js:226]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#Aside_%E2%80%94_.mjs_versus_.js

File: tests_integration/__tests__/esm-bundle.js
Comment (Monchi) [reply tests_integration/__tests__/esm-bundle.js]: Thank you for explanation.

It seems that current `parser-postcss.mjs` that built from umd exports only `parserPostcss`. No global `prettierPlugins`. I am missing something?

Comment (fisker) [reply tests_integration/__tests__/esm-bundle.js]: You are right, there is no `prettierPlugins` in `parser-postcss.mjs`.

I guess it's got tree-shaked?

**Review by fisker — APPROVED** [review]
Thank you for working on this, I think we are ready.

**Comment by Monchi** [comment]
@fisker Yes, we are ready. Thank you for your support!

**Comment by fisker** [comment]
Forgot to mention, I removed `doc.mjs`.

File: tests_integration/__tests__/esm-bundle.js
Comment (Monchi) [reply tests_integration/__tests__/esm-bundle.js]: I don't know the detailed algorithm of tree-shaking, but it seems there is nothing other than it.

**Comment by Monchi** [comment]
:+1: i saw the commit

**Review by sosukesuzuki — APPROVED** [review]
LGTM

**Review by evilebottnawi — APPROVED** [review]

**Pull request merged**
